### PR TITLE
Add ZenEx and message handlers

### PR DESCRIPTION
**Pull request by doomspork**
This is a port of the Zendesk functionality we have in HAL to a standalone queue powered service that allows us greater flexibility and easier to maintain chunks of functionality.

### CODE REVIEW

**Comment by doomspork** [comment]
I need to finish migrating the ticket creation to ZenEx as I did the others :+1:

**Comment by doomspork** [comment]
We also need to look at contributing back some improvements to ZenEx as we had discussed some time ago: https://github.com/otoyo/zen_ex/issues/16

File: test/test_helper.exs [hunk 1]
```diff
@@ -1 +1,16 @@
 ExUnit.start()
+
+Mox.defmock(HelpDesk.MockOrganizations, for: HelpDesk.Organizations)
+Mox.defmock(HelpDesk.MockUsers, for: HelpDesk.Users)
+Mox.defmock(HelpDesk.MockTickets, for: HelpDesk.Tickets)
+
+def HelpDesk.BroadwayHelpers do
```
Comment (btkostner) [inline test/test_helper.exs]: Small nitpick but this should probably go into the test support folder.

**Review by btkostner — APPROVED** [review]
Looks like mostly a copy of the current code we have in hal 🎊 

Question about deployment: I see we have the `apply_macro` code in here. Are we planning to setup plug and have this app handle the zendesk webhook, or do you want hal to proxy it?

**Comment by doomspork** [comment]
> Question about deployment: I see we have the apply_macro code in here. Are we planning to setup plug and have this app handle the zendesk webhook, or do you want hal to proxy it?

HAL should proxy the third party webhooks. We could create a generate "webhook" event to pass along any and all that occur, allowing the service to decide which to implement and which to ignore.

**Comment by btkostner** [comment]
That seems a little weird. What microservice would listen to webhook events happening to other services? Wouldn't it be better if the zendesk service was the only one listening to zendesk webhooks, and if another service needed information, helpdesk would just put a bottle message into a queue?

That would keep a clear distinction between what is internal data and external service data, and would keep a more strict one microservice to one third party service architecture.

**Comment by doomspork** [comment]
Sorry I wasn't clear. Yes Zendesk would listen to only Zendesk webhooks. I was referring to allowing the service to decide which events from that provider it wanted to listen to; Zendesk sends us a lot more than `apply_macros` but we ignore a lot and that should happen in the service not the gateway.

**Comment by doomspork** [comment]
@btkostner thoughts on what a generic webhook event could be? `Callbacked`, `Webhooked`, `WebhookReceived`?

**Comment by btkostner** [comment]
Personally, I still think that this service should be outside of our gateway considering only a single service should be sending messages to it, and it handles no JWT auth. Trying to handle each services authentication / verification logic in the gateway is going to be pretty messy. Trying to figure out an agnostic way to pass the request in a queue so each service can authenticate and verify itself is going to be a lot of pointless work. And if we just have Hal proxy the whole request, might as well cut out the middle man and single point of failure.

What webhook are you talking about? This is what I thought you meant from above:
```
Zendesk > Sends webhook > Hal receives > Hal proxies > Help Desk receives Proxied request

Help Desk processes webhook and converts event to regular Bottle messages like `UserUpdated`, `OrganizationJoined`, etc.
```

If you still want to have Hal gateway this and put the message in a queue, I think the second easiest way is going to be having hal handle the auth and verification, then passing info in a message like so:

```ex
%Bottle.Core.V1.Webhook{
  service: "zendesk",
  request_id: "blablabla",
  message: "a request body string here"
}
```

Doing this with SNS we can add a `service` attribute when we submit it, and have AWS filter on that to ensure only the service that handles that message gets that message.

*If* you wanted to do authentication / verification on the service layer, we would need to keep in mind all the different ways services do that. Stripe sends a signature header. Mailchimp has no built in way so we just add a string it the end of the url. Other services might have _fun new creative ways_. So we would probably need to add both of those fields.

```ex
%Bottle.Core.V1.Webhook{
  service: "zendesk",
  request_id: "blablabla",
  headers: %{
    authentication: "blabla"
  },
  path: "/webohook/asdf/blablabla",
  message: "a request body string here"
}
```

**Comment by doomspork** [comment]
> Personally, I still think that this service should be outside of our gateway considering only a single service should be sending messages to it, and it handles no JWT auth. Trying to handle each services authentication / verification logic in the gateway is going to be pretty messy.

I can certainly change my thinking but generally speaking making a service that has no reason to be publicly accessible public just opens up an attack vector into the platform. Especially since this service has access to our database(s) and queue(s). Whereas a API gateway would have very _very_ minimal access to infrastructure; we aren't there yet because of the legacy functionality we've still got to extract from HAL but at some point HAL would no longer need database access.

We'd also need to make our Mailchimp service and Strip services :exclamation: publicly accessible for their webhooks. The latter alone makes me sweat a little. Given the opportunity to hide my payment service far from the public internet I'd be hard pressed to pass up on that opportunity.

>  Trying to figure out an agnostic way to pass the request in a queue so each service can authenticate and verify itself is going to be a lot of pointless work. And if we just have Hal proxy the whole request, might as well cut out the middle man and single point of failure.

We terminate JWT in HAL, why would we consider the auth for these services an exception? I see very little reason we couldn't enforce the arguably very simple token authentication we use for most of our webhooks in HAL. At which point all you're passing to the service is all the authenticated webhooks, letting them decide which to implement.

> What webhook are you talking about? This is what I thought you meant from above:

Right now we have three for Zendesk. For Stripe we have a half dozen. We listen to two Mailchimp webhooks but arguably should be listening to more.

What I was suggesting is:

1. API Gateway authenticates the webhook request.
2. Omits a `WebhookReceived` event to the corresponding service.
3. Service decide whether or not to take action for a webhook or whether to ignore. 

We actually have some examples in HAL that I think will help clarify this :tada: 

If we peek at our Mailchimp Webhook controller it implements auth and decides which events from Mailchimp we listen to and which we ignore: https://github.com/system76/hal/blob/master/lib/hal_web/controllers/webhooks/mailchimp_controller.ex#L11

However if we peek at how we did Zendesk's Webhook controller it implements auth and then passes off the event to be decided by the domain logic:
https://github.com/system76/hal/blob/master/lib/hal_web/controllers/webhooks/zendesk_controller.ex#L37

Where we decide how to handle them:
https://github.com/system76/hal/blob/master/lib/zendesk/events.ex#L13-L39

I'm suggesting taking the latter approach and continuing to use it for microservices. The controller will send a `WebhookReceived` event for the service to listen to and act on. 

Does that help?

**Comment by btkostner** [comment]
Much. In that case, I'd do the second code block from above.

**Comment by doomspork** [comment]
I am going to merge this and come back to fixing up the webhook stuff since we're not going to be using it during this initial roll out.

**Pull request merged**